src/acl/ParameterizedNode.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_ACL_PARAMETERIZEDNODE_H
+#define SQUID_SRC_ACL_PARAMETERIZEDNODE_H
+
+#include "acl/Acl.h"
+#include "acl/Data.h"
+#include "base/Assure.h"
+
+#include <memory>
+
+namespace Acl
+{
+
+/// An ACL that manages configured parameters using a given Parameters class P.
+/// That Parameters class must support an ACLData<> or equivalent API.
+template <class P>
+class ParameterizedNode: public ACL
```

Comment:
> I am finding it very hard so far to tell what this template is actually supposed to be doing. 

Sorry to hear that. FWIW, I believe the current "ACL that manages configured parameters using a given Parameters class" definition is a good one. I cannot improve it without knowing what causes the difficulty you are describing.

> It looks like you are just shuffling things around, but the renaming is obfuscating which parts are the strategy vs the config data.

* The "config data" is ParameterizedNode::data. It used to be ACLStrategized::data. No significant changes here.
* The "transaction data extraction and matching strategy" is ParameterizedNode::match(), which is our regular ACL::match(). It used to be ACLStrategized::matcher->match() which is _not_ a part of the ACL class/hierarchy, causing the problems described in the PR description and addressed in this PR.


> Also, what is the performance difference between these two techniques?

The new types remove one forwarding hop (from ACLStrategised to ACLStrategised::matcher), so they ought to be a tiny bit faster, but this PR is not a performance optimization. The focus here is on solving serious architectural problems detailed in the PR description. We are lucky that these solutions do not hurt performance (AFAICT), but they would be necessary even if they did.

Reply:
> The new types remove one forwarding hop (from ACLStrategised to ACLStrategised::matcher), so they ought to be a tiny bit faster

When I wrote the above, I did not realize that Strategy-derived classes are not pooled at all in the official code! With this PR changes, that problem is gone (with the Strategy template class itself) and all formerly ACLStrategised classes get individual memory pools, typically one per acltype, with proper acltype-based naming in mgr:mem reports. That improves reporting and may help with memory leak triage while pooling might speed up reconfiguration a little.

Still, this PR is not really a performance optimization. The focus here is on solving serious architectural problems detailed in the PR description. We are lucky that these solutions do not hurt performance (AFAICT), but they would be necessary even if they did.